File: DS_Problems/Graph/kosaraju.cpp
/*
Steps:

1. Sort all the nodes in order of finishing time -> Toposort
2. Transpose the graph
3. DFS traversal according to the finishing time
*/


/*
Time complexity -> O(N+E)
Space Complexity -> O(N+E)+O(N)+O(N)
*/


#include<bits/stdc++.h>
using namespace std;


void findTopoSort( vector<int> &vis ,int node, stack<int> &st, vector<int> adj[]){
    vis[node] = 1;
    for(auto it: adj[node]){
        if(!vis[it]){
            findTopoSort(vis, it, st, adj );
        }
    }

    st.push(node);
}


void revDFS(int node, vector<int> &vis, vector<int> transpose[]){
    cout<<node<<" ";
    vis[node] = 1;
    for(auto it: transpose[node]){
        if(!vis[it]){
            revDFS(it, vis, transpose);
        }
    }
}

void findSCC(int n, vector<int> adj[]){ // Function for finding strongly connected components
    vector<int> vis(n+1, 0);
    stack<int> st;

    for(int i=1;i<=n;i++){ // First step
        if(!vis[i]){
            findTopoSort(vis, i, st, adj);
        }
    }

    // Second step

    vector<int> transpose[n+1];
    
    for(int i=1;i<=n;i++){
        vis[i] = 0;
        for(auto it: adj[i]) {
            transpose[it].push_back(i);
        }
    }

    // Third step

    while(!st.empty()){
        int node = st.top();
        st.pop();
        if(!vis[node]){
            cout<<"SCC: ";
            revDFS(node, vis, transpose);
            cout<<endl;
        }
    }

}

void adjList(int n, int m){
    vector<int> adj[n+1];
    
    for(int i=1;i<=m;i++){
        int u, v;
        cout<<"Enter two nodes: "<<endl;
        cin >> u >> v;
        adj[u].push_back(v);
    }

    findSCC(n, adj);
}

int main(){
    int n, m;
    cin >> n >> m;
    adjList(n, m);
}

File: DS_Problems/Graph/Kosaraju.java
package DS_Problems.Graph;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Stack;

/*
Steps:

1. Sort all the nodes in order of finishing time -> Toposort
2. Transpose the graph
3. DFS traversal according to the finishing time
*/

/*
Time complexity -> O(N+E)
Space Complexity -> O(N+E)+O(N)+O(N)
*/

public class Kosaraju {
    static Map<Integer, List<Integer>> graph = new HashMap<>();

    public static void main(String[] args) {
        int[][] edges = {
                { 3, 2 },
                { 2, 1 },
                { 1, 3 },
                { 2, 4 },
        };

        for (int i = 0; i < edges.length; i++) {
            int from = edges[i][0];
            int to = edges[i][1];
            List<Integer> newItemsForFrom = graph.getOrDefault(from, new ArrayList<>());
            newItemsForFrom.add(to);
            graph.put(from, newItemsForFrom);
        }

        int nodes = 4;
        findStronglyConnectedComponents(nodes);
    }

    public static void findStronglyConnectedComponents(int nodes) {
        boolean[] vis = new boolean[nodes + 1];
        Stack<Integer> s = new Stack<>();
        for (int node = 1; node <= nodes; node++) {
            if (!vis[node]) {
                topoSort(node, vis, s);
            }
        }
        Map<Integer, List<Integer>> transpose = new HashMap<>();

        for (int node = 1; node <= nodes; node++) {
            vis[node] = false;
            for (Integer neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                List<Integer> newItems = transpose.getOrDefault(neighbor, new ArrayList<>());
                newItems.add(node);
                transpose.put(neighbor, newItems);
            }
        }

        while (!s.isEmpty()) {
            int node = s.pop();
            if (!vis[node]) {
                System.out.print("SCC: ");
                reverseDFS(node, vis, transpose);
                System.out.println("");
            }
        }

    }

    private static void topoSort(int node, boolean[] vis, Stack<Integer> s) {
        vis[node] = true;

        for (Integer neighbor : graph.getOrDefault(node, new ArrayList<>())) {
            if (!vis[neighbor]) {
                topoSort(neighbor, vis, s);
            }
        }

        s.add(node);
    }

    private static void reverseDFS(int node, boolean[] vis, Map<Integer, List<Integer>> transpose) {
        vis[node] = true;
        System.out.print(node + " ");

        for (Integer neighbor : transpose.getOrDefault(node, new ArrayList<>())) {
            if (!vis[neighbor]) {
                reverseDFS(neighbor, vis, transpose);
            }
        }
    }
}
